subprojects/docs/src/samples/build-organization/cross-project-output-sharing/kotlin/producer/build.gradle.kts

```
@@ -0,0 +1,17 @@
+val sharedFile = project.layout.buildDirectory.file("some-subdir/shared-file.txt")
+
+val makeFile = tasks.register("makeFile") {
+    outputs.file(sharedFile)
+    doFirst {
+        sharedFile.get().asFile.writeText("This file is shared across Gradle subprojects.")
+    }
+}
+
+val sharedConfiguration by configurations.creating {
+    isCanBeConsumed = true
+    isCanBeResolved = false
+}
+
+artifacts {
+    add(sharedConfiguration.name, sharedFile)
```

This should use the task and not the simple reference to the produced file. The Groovy samples does the right thing already.

```suggestion
    add(sharedConfiguration.name, makeFile)
```

---

I tested this locally here, and found that referencing the task in Kotlin doesn't work. Maybe needs a little more love, using builtBy or similar. I'll have a go now.

---

(The failure is like this):

```
robmoore@Robs-MacBook-Pro gradle-share-outputs-between-projects % gw :c:showFile
Unable to find a gradle build file named build.gradle or build.gradle.kts.
Using gradle at '/Users/robmoore/IdeaProjects/gradle-share-outputs-between-projects/gradlew' to run buildfile '':

Configuration cache is an incubating feature.
Calculating task graph as no configuration cache is available for tasks: :c:showFile

> Task :consumer:showFile
File is at /Users/robmoore/IdeaProjects/gradle-share-outputs-between-projects/producer/build/some-subdir/shared-file.txt

FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache problems found in this build.

1 problem was found storing the configuration cache.
- Task `:producer:makeFile` of type `org.gradle.api.DefaultTask`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.4.2/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
```

---

The reported problem is not about that usage but about the fact that the `val sharedFile`, declared at the root of the script, and used in the task action, captures the script context and thus cannot be serialized by the configuration cache.

Here's the complete producer script that works:

```kotlin
val makeFile = tasks.register("makeFile") {
    val sharedFile = project.layout.buildDirectory.file("some-subdir/shared-file.txt")
    outputs.file(sharedFile)
    doFirst {
        sharedFile.get().asFile.writeText("This file is shared across Gradle subprojects.")
    }
}

val sharedConfiguration by configurations.creating {
    isCanBeConsumed = true
    isCanBeResolved = false
}

artifacts {
    add(sharedConfiguration.name, makeFile)
}
```

---

Thanks! Nice one

---

Applying now 👍

---

This highlights the fact that the sample is kind of broken!

When using the simple reference to the file the producer task was not run. But it was not noticeable because the consuming task only displays the path to the file. It should display the file's content! Then it would have been obvious that the script was wrong.

---

Ah right, of course. I'll change it to show the file's content.